File: webapp/src/app/_services/data/institution.service.ts
import { Injectable } from '@angular/core';
import {Observable} from 'rxjs';
import {ResponseObject} from "../../types/ResponseObject";
import {Page} from "../../types/page";
import {HttpClientService} from "../http-client-service";
import {Institution} from "../../types/Institution";
import {PAGE_SIZE} from "./CONST_VAR";



@Injectable({
    providedIn: 'root',
})
export class InstitutionService {
    private controller = '/institutions';

    constructor( private http: HttpClientService) {
    }
    getInstitutions(query= '', page = 0, pageSize = PAGE_SIZE, sortDirection = 'name'): Observable<ResponseObject<Page<Institution>>>{

        return this.http.get<ResponseObject<Page<Institution>>>('/institutions' +
            `?query=${query}&page=${page}&size=${pageSize}&sortDirection=${sortDirection}`);

    }
    getInstitutionsWithQuery(query: string, page = 0, pageSize= PAGE_SIZE, sortDirection= 'firstName'):
        Observable<ResponseObject<Page<Institution>>>  {
// TODO implement logic on Server
        return this.http.get<ResponseObject<Page<Institution>>>(
            `/institutions/search?query=${query}&page=${page}&size=${pageSize}&sortDirection=${sortDirection}`);
    }
    getInstitution(id: number): Observable<ResponseObject<Institution>> {
        return this.http.get<ResponseObject<Institution>>(`${this.controller}/${id}`);
    }
    createInstitution(data:any): Observable<ResponseObject<any>> {
        return this.http.post<ResponseObject<any>>(this.controller, data );
    }

    updateInstitution(data: any): Observable<ResponseObject<Institution>> {
        return this.http.put<ResponseObject<Institution>>(this.controller ,data);
    }


    deleteInstitution(id: number) {
        return this.http.delete<ResponseObject<any>>(`${this.controller}/${id}`);
    }
}

File: server/src/main/java/com/school_system/service/InstitutionService.java
package com.school_system.service;

import com.school_system.common.ResponseObject;
import com.school_system.dto.request.InstitutionRequest;
import com.school_system.dto.response.InstitutionResponse;
import com.school_system.entity.school.Institution;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.transaction.annotation.Transactional;

public interface InstitutionService {
    ResponseObject<InstitutionResponse> createInstitution(InstitutionRequest institutionRequest);

    ResponseObject<InstitutionResponse> getInstitution(Long id);

    ResponseObject<Page<InstitutionResponse>> getInstitutions(Pageable pageable);

    ResponseObject<InstitutionResponse> updateInstitution(InstitutionRequest institutionRequest);

    ResponseObject<Page<InstitutionResponse>> getInstitutionsWithQuery(String query, Pageable pageable);

    ResponseObject<String> deleteInstitution(Long id);
}
